packages/ast/src/lowering.candy

```diff
@@ -0,0 +1,339 @@
+use compiler_utils
+use cst
+use ..declarations
+use ..errors
+use ..expressions
+use ..file
+use ..node
+
+public fun lowerCstToAst(
+  file: File,
+  cst: CstNode<CstFile>,
+): (AstFile, Map<CstNodeId, AstNodeId>, List<CompilerError>) {
+  let context = LoweringContext.create(file)
+  let ast = lowerFile(context, cst)
+  let trackingResult = context.getResult()
+  Tuple(ast, trackingResult.first, trackingResult.second)
+}
+
+// file
+
+fun lowerFile(context: LoweringContext, cst: CstNode<CstFile>): AstFile {
+  let id = AstFileId(context.file)
+
+  mut let sawDeclaration = false
+  let useLines = MutableList.empty<AstUseLine>()
+  let declarations = MutableList.empty<AstModule | AstTrait | AstImpl | AstClass | AstFunction | AstProperty>()
+  for node in cst.child.content {
+    if node.child is CstUseLine {
+      if sawDeclaration {
+        context.addError(
+          UseLineAfterDeclarationCompilerError(context.file, node.cast<CstUseLine>()),
+        )
+      }
+
+      let result = lowerUseLine(
+        context,
+        id,
+        node.cast<CstUseLine>(),
+        (useLines as Iterable<AstUseLine>).length(),
+      )
+      if result is Some<AstUseLine> {
+        useLines.append(result.unwrap())
+      }
+    } else {
+      sawDeclaration = true
+    }
+  }
+
+  context.addIdMapping(cst.id, id)
+  AstFile(id, useLines, declarations)
+}
+
+fun lowerUseLine(
+  context: LoweringContext,
+  fileId: AstFileId,
+  cst: CstNode<CstUseLine>,
+  useLineIndex: Int,
+): Maybe<AstUseLine> {
+  let id = AstUseLineId(fileId, useLineIndex)
+
+  let modifiers = lowerModifiers(context, id, cst.child.modifiers)
+  let target = cst.child.target
+      .map<AstUseLineTarget>({
+        if it.child is CstLocalRelativeUseLineTarget {
+          lowerLocalRelativeUseLineTarget(context, id, it.cast<CstLocalRelativeUseLineTarget>())
+        } else {
+          lowerGlobalUseLineTarget(context, id, it.cast<CstGlobalUseLineTarget>())
+        }
```

Comment:
Fine for now. In the future, instead of many top-level functions, we should probably have a `LowerToAst` trait or something similar that is implemented for `CstNode<CstFile>`, `Cst<Node<CstUseLine>>` etc. so we can just to `it.toAst()`